server/lib/src/plugins/namehistory.rs

```diff
@@ -0,0 +1,223 @@
+        let mut set = HashSet::new();
+        set.insert(PartialValue::new_iutf8("account"));
+        set
+    };
+    static ref HISTORY_ATTRIBUTES: [&'static str;1] = ["name"];
+}
+
+impl NameHistory {
+    fn handle_name_updates(
+        pre_cand: &[Arc<EntrySealedCommitted>],
+        cand: &mut Vec<EntryInvalidCommitted>,
+        cid: &Cid,
+    ) -> Result<(), OperationError> {
+        for (pre, post) in pre_cand.iter().zip(cand) {
+            // here we check if the current entry has at least one of the classes we intend to target
+            // maybe using a filter would be more appropriate?
+            if CLASSES_TO_UPDATE
+                .iter()
+                .any(|pv| post.attribute_equality("class", pv))
+            {
+                for history_attr in HISTORY_ATTRIBUTES.iter() {
+                    let pre_name_option = pre.get_ava_single(history_attr);
+                    let post_name_option = post.get_ava_single(history_attr);
+                    if let (Some(pre_name), Some(post_name)) = (pre_name_option, post_name_option) {
+                        let ava_name = format!("{}_history", history_attr);
+                        if pre_name != post_name {
+                            //// WARNING!!! this match will have to be adjusted based on what kind of attribute
+                            //// we are matching on, for example for displayname we would have to use Value::utf8 instead!!
+                            // as of now we're interested just in the name so we use Iname
+                            match pre_name {
+                                Value::Iname(n) => post.add_ava_if_not_exist(
+                                    &ava_name,
+                                    Value::AuditLogString(cid.clone(), n),
+                                ),
+                                _ => return Err(OperationError::InvalidValueState),
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Plugin for NameHistory {
+    fn id() -> &'static str {
+        "plugin_name_update"
+    }
+
+    fn pre_modify(
+        qs: &mut QueryServerWriteTransaction,
+        pre_cand: &[Arc<EntrySealedCommitted>],
+        cand: &mut Vec<EntryInvalidCommitted>,
+        _me: &ModifyEvent,
+    ) -> Result<(), OperationError> {
+        Self::handle_name_updates(pre_cand, cand, qs.get_txn_cid())
+    }
+
+    fn pre_batch_modify(
+        qs: &mut QueryServerWriteTransaction,
+        pre_cand: &[Arc<EntrySealedCommitted>],
+        cand: &mut Vec<EntryInvalidCommitted>,
+        _me: &BatchModifyEvent,
+    ) -> Result<(), OperationError> {
+        Self::handle_name_updates(pre_cand, cand, qs.get_txn_cid())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use crate::entry::{Entry, EntryInit, EntryNew};
+    use crate::prelude::uuid;
+    use crate::repl::cid::Cid;
+    use crate::value::Value;
+
```

Comment:
I think the behaviour we want is that we want the name history to show the time when a name was *added* not the time when it was *removed* from the account. If you do it as removed, you'll hit replication issues.

So how would you re-model this part of the code so that when you create the entry, the cid of the create and name is added? When the modify shows a new name, add the new name to the history rather than the old one?